src/components/tl/ucp/allgather/allgather_neighbor.c

```diff
@@ -0,0 +1,176 @@
+#include "utils/ucc_math.h"
+#include "utils/ucc_coll_utils.h"
+#include "components/mc/ucc_mc.h"
+
+static ucc_rank_t get_recv_from_rank(ucc_rank_t rank, ucc_rank_t size, int i)
+{
+    const int  i_parity = i % 2;
+    ucc_rank_t neighbor[2], offset_at_step[2], recv_data_from[2];
+    if (rank % 2) {
+        neighbor[0]       = (rank - 1 + size) % size;
+        neighbor[1]       = (rank + 1) % size;
+        recv_data_from[0] = neighbor[0];
+        recv_data_from[1] = neighbor[0];
+        offset_at_step[0] = (-2);
+        offset_at_step[1] = (+2);
+    } else {
+        neighbor[0]       = (rank + 1) % size;
+        neighbor[1]       = (rank - 1 + size) % size;
+        recv_data_from[0] = rank;
+        recv_data_from[1] = rank;
+        offset_at_step[0] = (+2);
+        offset_at_step[1] = (-2);
+    }
+
+    return (recv_data_from[i_parity] +
+            offset_at_step[i_parity] * ucc_div_round_up(i, 2) + size) %
+           size;
+}
+
+ucc_status_t ucc_tl_ucp_allgather_neighbor_init(ucc_base_coll_args_t *coll_args,
+                                                ucc_base_team_t      *team,
+                                                ucc_coll_task_t     **task_h)
+{
+    ucc_status_t       status = UCC_OK;
+    ucc_tl_ucp_task_t *task;
+    ucc_tl_ucp_team_t *ucp_team;
+
+    task     = ucc_tl_ucp_init_task(coll_args, team);
+    ucp_team = TASK_TEAM(task);
+
+    if (!ucc_coll_args_is_predefined_dt(&TASK_ARGS(task), UCC_RANK_INVALID)) {
+        tl_error(UCC_TASK_LIB(task), "user defined datatype is not supported");
+        status = UCC_ERR_NOT_SUPPORTED;
+        goto out;
+    }
+
+    if (UCC_TL_TEAM_SIZE(ucp_team) % 2) {
+        tl_warn(UCC_TASK_LIB(task),
+                "odd team size is not supported, switching to ring");
+        status = ucc_tl_ucp_allgather_ring_init_common(task);
+    } else {
+        task->super.post     = ucc_tl_ucp_allgather_neighbor_start;
+        task->super.progress = ucc_tl_ucp_allgather_neighbor_progress;
+    }
+
+out:
+    if (status != UCC_OK) {
+        ucc_tl_ucp_put_task(task);
+    }
+
+    *task_h = &task->super;
+    return UCC_OK;
+}
+
+/* Original implmenetation: https://github.com/open-mpi/ompi/blob/main/ompi/mca/coll/base/coll_base_allgather.c */
+void ucc_tl_ucp_allgather_neighbor_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task      = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team      = TASK_TEAM(task);
+    ucc_rank_t         trank     = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t         tsize     = UCC_TL_TEAM_SIZE(team);
+    void              *rbuf      = TASK_ARGS(task).dst.info.buffer;
+    ucc_memory_type_t  rmem      = TASK_ARGS(task).dst.info.mem_type;
+    ucc_datatype_t     dt        = TASK_ARGS(task).dst.info.datatype;
+    size_t             count     = TASK_ARGS(task).dst.info.count;
+    size_t             data_size = (count / tsize) * ucc_dt_size(dt);
+    ucc_rank_t         neighbor[2], i;
```

Comment:
perhaps call the array neighbors instead of neighbor?

Reply:
ok, I'm not native speaker :+1: